### PR TITLE
Use new sublayer querying api when adding layers

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR:
- Reworks all non-browser methods of adding layers to maps to use the new querySublayers api, giving a much improved experience when adding layers from mixed format files (such as geopackage)
- Ports a ton of logic which was previously in qgisapp.cpp and other app related classes to the relevant provider's querySublayer methods, and add tests for these
- Deprecates the old sublayers dialog class and introduces a new dialog based on the new api, which nicely handles mixed format sublayers. It also:
    - Resolves geometry types in a background thread, to avoid lengthy
application hangs whenever a full table scan is required to determine
available geometry types
    - Shows a handy search filter box for filtering sublayers to matching
strings
    - Shows icons representing sublayer types
    - Has an interactive label showing the file path which can be clicked
to open a file explorer focused on the file
- Removes a ton of near duplicate code
- Fixes a bunch of bugs along the way

(Thanks to a QGIS grant)

### CODE REVIEW

**Comment by nirvn** [comment]
@nyalldawson , holy moly this is exciting! Do you have a screenshot of the new dialog at-hand?

**Comment by nyalldawson** [comment]
@nirvn here you go:

Showing mixed layer geopackage with vector/raster:

image

Showing mixed mesh/raster layers from a .nc file:

image


Showing scanning in progress for a large mixed geometry type tab file:
image

When the scan finishes the dialog is immediately updated:

image

**Comment by nyalldawson** [comment]
(@nirvn A final follow up PR will also show project items inside geopackages in the dialog!)

**Comment by nirvn** [comment]
@nyalldawson , sweet. Are those (numbers) feature counts here:
image

If so, could we get thousand separators whenever the locale has those on?

**Comment by nyalldawson** [comment]
@nirvn 
>If so, could we get thousand separators whenever the locale has those on?

It already is (but I personally can't stand them so have disabled them)

**Comment by nirvn** [comment]
:bow:

File: src/app/qgisapp.cpp [hunk 1]
```diff
@@ -2043,7 +2046,7 @@ void QgisApp::handleDropUriList( const QgsMimeDataUtils::UriList &lst )
 
         for ( const QString &w : std::as_const( warnings ) )
         {
-          message += QStringLiteral( "<li>%1</li>" ).arg( w.toHtmlEscaped().replace( '\n', QLatin1String( "<br>" ) ) );
+          message += QStringLiteral( "<li>%1</l>" ).arg( w.toHtmlEscaped().replace( '\n', QLatin1String( "<br>" ) ) );
```
Comment (rouault) [inline src/app/qgisapp.cpp:2049]: unintented change I guess ?

Comment (nyalldawson) [reply src/app/qgisapp.cpp]: Yep, thanks

**Comment by PeterPetrik** [comment]
nice!

**Comment by nyalldawson** [comment]
Approved indirectly in https://github.com/qgis/QGIS/pull/44105

**Comment by nirvn** [comment]
@nyalldawson , I think the merging of this PR caused the following regression: I can't add an XYZ layer from the browser anymore. I now get an invalid data source error message bar triggered here: https://github.com/qgis/QGIS/blob/master/src/app/qgisapp.cpp#L13079

**Comment by elpaso** [comment]
@nyalldawson I cannot add any PG layer too:  I think that `::querySublayers` must be implemented for all providers or the logic needs to be changed in https://github.com/qgis/QGIS/blob/master/src/app/qgisapp.cpp#L13094

**Comment by vcloarec** [comment]
@nyalldawson , same problem with MSSQL, from this commit:
https://github.com/qgis/QGIS/pull/44085/commits/2ad0aca3326cc4ed2c5c0852d79ca0c50634206e

**Comment by nyalldawson** [comment]
@vcloarec it should be fixed in master already

**Comment by vcloarec** [comment]
@nyalldawson ah .... sorry for the noise... my master was 2 days...

**Comment by uclaros** [comment]
> @nyalldawson , sweet. Are those (numbers) feature counts here:
> image
> 
> If so, could we get thousand separators whenever the locale has those on?

@nyalldawson Is there any particular reason why you added the feature count in parentheses in the description field instead of a separate sortable field with a header like it was in the old UI?
I mean, if a dev has to ask then it will probably not be obvious to some users...

**Comment by nyalldawson** [comment]
@uclaros 
It's because the new dialog shows all layer types, so I was wanting to keep the columns generic and not have layer-type specific columns which will be empty for some layer types.

**Comment by uclaros** [comment]
@nyalldawson OK then, it's no big deal!
I miss however having all layers pre-selected and being able to sort by geometry type (to easily add all point layers for example). Do you have something against those two?

**Comment by nyalldawson** [comment]
@uclaros addressed in https://github.com/qgis/QGIS/pull/44572

**Pull request merged**